src/Lights/Shadows/shadowGenerator.ts

```diff
@@ -656,10 +681,100 @@ export class ShadowGenerator implements IShadowGenerator {
         if (this._shadowMap2) {
             return this._shadowMap2;
         }
-
         return this._shadowMap;
     }
 
+    private _viewSpaceFrustums: Array<Array<Vector3>>;
+    private _viewSpaceBoundingSpheres: Array<BoundingSphere>;
+    private _CSMtransformMatrices: Array<Matrix>;
+    private _cascades = 1;
+    /**
+     * Gets the number of cascades used in CSM mode.
+     */
+    public get cascades(): number {
+        return this._cascades;
+    }
+    /**
+     * Sets the number of cacades used in CSM mode.
+	 * Limited to 2 to 4 cascade to reduce resource footprint
+     */
+    public set cascades(value: number) {
+        if (this._cascades === value || value < ShadowGenerator.MIN_CASCADES || value > ShadowGenerator.MAX_CASCADES) {
+            return;
+        }
+        this._cascades = value;
+        if (this._cascades >= ShadowGenerator.MIN_CASCADES && this._cascades <= ShadowGenerator.MAX_CASCADES) {
+            this.useCSM = true;
+        }
+    }
+
+    private _useCSM = false;
+    /**
+     * Gets the CSM mode.
+     */
+    public get useCSM(): boolean {
+        return this._useCSM;
+    }
+    /**
+     * Sets the CSM mode.
+     */
+    public set useCSM(value: boolean) {
+        if (this._useCSM === value) {
+            return;
+        }
+        this._useCSM = value;

```

Comment:
this should recreate the internal textures/post pro like the other modes I guess

Reply:
Are you suggesting the user implement csm with smthg like 
```
var shadowgenerator = new ShadowGenerator(512, light);
shadowgenerator.useCSM = true;
```
This was my initial take but I found it inefficient. Reason is in the comments on the constructor.

Reply:
Agree so you can keep the ctor params but we still need to handle the dynamic case in case the configuration changes later on.

Reply:
ok, added an init and dispose checks.